pkg/grafana/client.go

```diff
@@ -0,0 +1,101 @@
+package grafana
+
+import (
+	"net/url"
+	"strings"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	log "github.com/sirupsen/logrus"
+)
+
+type Client struct {
+	client GAPIClient
+}
+
+type GAPIClient interface {
+	UserByEmail(email string) (user gapi.User, err error)
+	CreateUser(user gapi.User) (int64, error)
+	AddOrgUser(orgID int64, user, role string) error
+}
+
+func NewClient(baseURL *url.URL, cfg gapi.Config) (*Client, error) {
+	newClient := &Client{}
+
+	client, err := gapi.New(baseURL.String(), cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	newClient.client = client
+
+	return newClient, nil
+}
+
+// LookupUser search for a user by Login or Email and returns it
+func (c *Client) LookupUser(loginOrEmail string) (*gapi.User, error) {
+	user, err := c.client.UserByEmail(loginOrEmail)
+	if err != nil {
+		return nil, err
+	}
+
+	// gapi returns an empty struct
+	if user.Login == "" {
+		return nil, err
+	}
+
+	return &user, nil
+}
+
+// CreateUser adds a new global user to Grafana
+func (c *Client) CreateUser(user gapi.User) (int64, error) {
+	uid, err := c.client.CreateUser(user)
+	if err != nil {
+		return uid, err
+	}
+
+	return uid, nil
+}
+
+// AddOrgUser adds a user, with a role, to an Organization specified by OrgID
+func (c *Client) AddOrgUser(OrgID int64, login string, role string) error {
+	err := c.client.AddOrgUser(OrgID, login, role)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpsertOrgUser adds a user to an Organization, creating it, if it doesn't exists
+func (c *Client) UpsertOrgUser(OrgId int64, user gapi.User, role string) error {
+	foundUser, err := c.LookupUser(user.Login)
+	if err != nil {
+		if !strings.Contains(err.Error(), "User not found") {
+			return err
+		}
+	}
+
+	if foundUser == nil {
+		log.Infof("no user with login %s found, creating new one", user.Login)
+		uid, err := c.CreateUser(gapi.User{
+			Login:    user.Login,
+			Email:    user.Email,
+			Password: "abc123",
```

Comment:
It's here to satisfy the API. I'll generate a secure password but it would be essentially the same for all new users.

As how the password can be used, we are essentially going to disable the local login mechanisms of Grafana as we serve it to end users, using Ingress I mean. I can test if I can use those passwords when bypassing Ingress, basically not using the auth proxy.

Reply:
Then its probably best practice to to generate a secure random string and leave it at that.

Reply:
done